src/web/TextInput.tsx

```diff
@@ -43,6 +43,72 @@ export interface TextInputContext {
     focusArbitrator?: FocusArbitratorProvider;
 }
 
+class TextInputPlaceholderSupport {
+    private _internalId = TextInputPlaceholderSupport.sequenceId();
+    private _styleElement: HTMLStyleElement | null = null;
+
+    private static sequenceId = (() => {
+        let id = 0;
+        return () => ++id;
+    })();
+
+    className(): string {
+        return `reactxp-text-input-${this._internalId}`;
+    }
+
+    tearDown() {
+        const styleElement = this._styleElement;
+        if (styleElement && styleElement.parentNode) {
+            styleElement.parentNode.removeChild(styleElement);
+        }
+    }
+
+    private _setupStyleElement(): HTMLStyleElement {
+        let style = this._styleElement;
+
+        if (!style) {
+            style = document.createElement('style');
+            style.type = 'text/css';
+            this._styleElement = style;
+        }
+
+        if (!style.parentNode) {
+            document.head.appendChild(style);
+        }
+
+        return style;
+    }
+
+    setColor(placeholderColor: string | undefined) {
+        if (typeof document === undefined) {
+            return;
+        }
+
+        if (placeholderColor) {
+            const selectors = [
+                '::placeholder', // Modern browsers
+                '::-webkit-input-placeholder',  // Webkit
+                '::-moz-placeholder', // Firefox 19+
+                ':-moz-placeholder', // Firefox 18-
+                ':-ms-input-placeholder' // IE 10+
+            ];
+            const className = this.className();
+            const cssStyle = selectors
+                .map((pseudoSelector) =>
+                    `.${className}${pseudoSelector} {\n` +
+                        `\topacity: 1;\n` +
```

Comment:
`opacity: 1` is used because some obscure old Firefox with usage of 0.2% applies custom opacity to placeholder. I enforce the opacity in favour of managing it using `rgba`